### PR TITLE
Reputation dropdown

### PR DESCRIPTION
**Pull request by kntsoriano**
## Description
This adds the reputation dropdown. The reputation count is preloaded on page load, as well as the icon signifying whether there are unread reputation tokens.

I've also added a temporary `ReputationsChannel` that we use to listen to websocket events. We refetch the whole list whenever a message is received through this channel.

## Questions
1. When do we mark a token as seen?

### CODE REVIEW

File: app/controllers/api/reputation_controller.rb [hunk 1]
```diff
@@ -9,7 +9,13 @@ def index
 
       render json: SerializePaginatedCollection.(
         tokens,
-        data: tokens.map(&:rendering_data)
+        data: tokens.map(&:rendering_data),
+        meta: {
+          links: {
+            tokens: Exercism::Routes.reputation_journey_url
+          },
+          reputation_sum: current_user.reputation
```
Comment (ErikSchierboom) [inline app/controllers/api/reputation_controller.rb]: We've used the "total" string to describe the total amount of things. Maybe do:

```suggestion
          total_reputation: current_user.reputation
```

File: app/javascript/components/dropdowns/Reputation.tsx [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  results: ReputationToken[]
+  meta: {
+    links: {
+      tokens: string
+    }
+    reputationSum: number
+  }
+}
+
+const DEFAULT_ERROR = new Error('Unable to reputation tokens')
```
Comment (ErikSchierboom) [inline app/javascript/components/dropdowns/Reputation.tsx]: I think there is a word missing:

```suggestion
const DEFAULT_ERROR = new Error('Unable to retrieve reputation tokens')
```

**Review by ErikSchierboom — APPROVED** [review]

**Comment by iHiD** [comment]
> When do we mark a token as seen?

When an element is focussed or hovered.

We might decide this doesn't work though, so it might change during beta testing.

**Comment by iHiD** [comment]
@kntsoriano I've pushed a few tweaks :) Could you check my changes please, and add the pinging to the API when the reputation should be marked as seen (on hover or focus).

**Comment by iHiD** [comment]
For some reason I don't understand, CI will just not run on this PR 🤷

**Comment by kntsoriano** [comment]
@iHiD Running now :) I've added the capability to mark it as seen. Could you check if you're happy with the changes I've done to the API?

File: app/controllers/api/reputation_controller.rb [hunk 3]
```diff
@@ -7,15 +7,17 @@ def index
...
       render json: SerializePaginatedCollection.(
         tokens,
-        data: tokens.map do |token|
-          token.rendering_data.merge({
-                                       links: {
-                                         mark_as_seen: Exercism::Routes.mark_as_seen_api_reputation_url(token.uuid)
-                                       }
-                                     })
-        end,
+        data: data,
```
Comment (iHiD) [inline app/controllers/api/reputation_controller.rb:20]: @kntsoriano Removing the hash here makes Rubocop not do stupid indenting :)

File: app/javascript/components/dropdowns/reputation/ReputationIcon.tsx [hunk 4]
```diff
@@ -23,7 +23,7 @@ export const ReputationIcon = forwardRef<
         <Icon icon="reputation" alt="Reputation" />
         <span>{reputation}</span>
       </div>
-      {isSeen ? null : <div className="--notification" />}
+      <div className={'--notification ' + (isSeen ? '' : 'unseen')} />
```
Comment (iHiD) [inline app/javascript/components/dropdowns/reputation/ReputationIcon.tsx:26]: @kntsoriano I've changed this so it doesn't replace the element, just the class, which means we can use a CSS transition

**Comment by iHiD** [comment]
@kntsoriano All happy and merging. JS tests are broken but I think just the normal unrelated failures we keep getting.

**Pull request merged**